flang/lib/Optimizer/Builder/IntrinsicCall.cpp

```diff
@@ -2574,6 +2578,26 @@ mlir::Value IntrinsicLibrary::genAtanpi(mlir::Type resultType,
   return builder.create<mlir::arith::MulFOp>(loc, atan, factor);
 }
 
+static mlir::Value genAtomBinOp(fir::FirOpBuilder &builder, mlir::Location &loc,
+                                mlir::LLVM::AtomicBinOp binOp, mlir::Value arg0,
+                                mlir::Value arg1) {
+  auto llvmPointerType = mlir::LLVM::LLVMPointerType::get(builder.getContext());
```

Comment:
Out of curiosity, do you now if FIR alias analysis is clever enough to go through convert from this LLVM type and deduce what the `llvm.atomicrmw` is reading/writing not (or mainly, to deduce what it is not reading/writing to)?

It makes sense to me that we interoperate with LLVM pointer type to avoid redefining all the intrinsics at the FIR level, I just wonder if we need to improve AA here.

Reply:
We always skip through `fir.convert`
```
        .Case<fir::ConvertOp>([&](auto op) {
          // Skip ConvertOp's and track further through the operand.
          v = op->getOperand(0);
          defOp = v.getDefiningOp();
        })
...
```
I was trying to make sense of the MemoryEffects of `LLVM_AtomicRMWOp` but couldn't quite make sense of it:
```
def LLVM_AtomicRMWOp : LLVM_MemAccessOpBase<"atomicrmw", [
      TypesMatchWith<"result #0 and operand #1 have the same type",
                     "val", "res", "$_self">]> {
```

We talked about these ops in the Lorado office hours today and while we need the hooks in place to create the appropriate instructions we may eventually replace them with higher level atomic operations. TDB.

Reply:
Right, the effect description is not straightforward. If found this patch that explains some of it: https://github.com/llvm/llvm-project/commit/feb7beaf70bace1c3ffafc7f732c8fadca5e8c9d

From what I understand, it is giving atomic operation a global write effects because the synchronization aspect (e.g., could write some global lock).